File: src/types/notification.ts
import type { AgentType } from './workspace'
import type { AgentStatus } from './plugin'

export type NotificationType =
  | 'info'
  | 'warning'
  | 'error'
  | 'success'
  | 'needs_input'
  | 'task_complete'
  | 'task_error'

export type NotificationPriority = 'low' | 'normal' | 'high' | 'urgent'

export interface NotificationInputRequest {
  requestId: string
  prompt: string
  options?: string[]
  responding: boolean
  response?: string
}

export interface Notification {
  id: string
  type: NotificationType
  priority: NotificationPriority
  title: string
  message: string
  timestamp: number
  read: boolean
  dismissed: boolean
  source: string
  agentType?: AgentType
  spaceId?: string
  paneId?: string
  inputRequest?: NotificationInputRequest
  actions?: NotificationAction[]
}

export interface NotificationAction {
  id: string
  label: string
  style: 'primary' | 'secondary' | 'danger'
}

export interface AgentStatusEntry {
  id: string
  name: string
  agentType: AgentType
  status: AgentStatus
  paneId?: string
  spaceId?: string
  lastAction: string
  lastActionAt: number
  connectedAt: number
  progress?: { current: number; total: number; label: string }
}

export type PluginStatus = 'active' | 'inactive' | 'error' | 'installing' | 'updating'

export interface Plugin {
  id: string
  name: string
  description: string
  version: string
  author: string
  status: PluginStatus
  enabled: boolean
  icon?: string
  config?: Record<string, unknown>
  installedAt: number
  updatedAt: number
  error?: string
  agentCount: number
  capabilities: string[]
}

File: frontend/src/types/notification.rs
use serde::{Deserialize, Serialize};
use strum::{Display, EnumString};

use super::plugin::AgentStatus;
use super::workspace::AgentType;

/// Kind of notification.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, EnumString, Display)]
#[strum(serialize_all = "snake_case")]
pub enum NotificationType {
    Info,
    Warning,
    Error,
    Success,
    NeedsInput,
    TaskComplete,
    TaskError,
}

/// Priority level of a notification.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, EnumString, Display)]
#[strum(serialize_all = "lowercase")]
pub enum NotificationPriority {
    Low,
    Normal,
    High,
    Urgent,
}

/// An input request attached to a notification.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct NotificationInputRequest {
    pub request_id: String,
    pub prompt: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub options: Option<Vec<String>>,
    pub responding: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response: Option<String>,
}

/// Style hint for an action button.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, EnumString, Display)]
#[strum(serialize_all = "lowercase")]
pub enum NotificationActionStyle {
    Primary,
    Secondary,
    Danger,
}

/// A clickable action on a notification.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct NotificationAction {
    pub id: String,
    pub label: String,
    pub style: NotificationActionStyle,
}

/// A notification displayed in the UI.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Notification {
    pub id: String,
    #[serde(rename = "type")]
    pub notification_type: NotificationType,
    pub priority: NotificationPriority,
    pub title: String,
    pub message: String,
    pub timestamp: i64,
    pub read: bool,
    pub dismissed: bool,
    pub source: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_type: Option<AgentType>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub space_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pane_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input_request: Option<NotificationInputRequest>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub actions: Option<Vec<NotificationAction>>,
}

/// An entry tracking an agent's live status.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AgentStatusEntry {
    pub id: String,
    pub name: String,
    pub agent_type: AgentType,
    pub status: AgentStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pane_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub space_id: Option<String>,
    pub last_action: String,
    pub last_action_at: i64,
    pub connected_at: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub progress: Option<ProgressInfo>,
}

/// Progress information for an agent or task.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProgressInfo {
    pub current: usize,
    pub total: usize,
    pub label: String,
}
